src/utils/display-observer.js

```diff
@@ -0,0 +1,226 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {getServiceForDoc, registerServiceBuilderForDoc} from '../service';
+import {pushIfNotExist, removeItem} from './array';
+import {rethrowAsync} from '../log';
+
+const SERVICE_ID = 'DisplayObserver';
+
+/** @implements {Disposable} */
+export class DisplayObserver {
+  /**
+   * @param {!AmpDoc} ampdoc
+   */
+  constructor(ampdoc) {
+    const {win} = ampdoc;
+
+    /** @private @const {!IntersectionObserver} */
+    this.docObserver_ = new win.IntersectionObserver(
+      (e) => this.observed_(e, 0),
+      {
+        root: ampdoc.getBody(),
+        threshold: 0.001,
+      }
+    );
+
+    /** @private @const {!IntersectionObserver} */
+    this.viewportObserver_ = new win.IntersectionObserver(
+      (e) => this.observed_(e, 1),
+      {threshold: 0.001}
+    );
+
+    /** @private {boolean} */
+    this.isDocVisible_ = ampdoc.isVisible();
+
+    /** @private {?UnlistenDef} */
+    this.visibilityUnlisten_ = ampdoc.onVisibilityChanged(() => {
+      if (ampdoc.isVisible() !== this.isDocVisible_) {
+        this.isDocVisible_ = ampdoc.isVisible();
+        this.docVisibilityChanged_();
+      }
+    });
+
+    /** @private @const {!Map<!Element, !Array<function(boolean)>>} */
+    this.targetObserverMultimap_ = new Map();
+
+    /** @private @const {!Map<!Element, !Array<boolean>>} */
+    this.targetDisplayMap_ = new Map();
+  }
+
+  /** @override */
+  dispose() {
+    this.docObserver_.disconnect();
+    this.viewportObserver_.disconnect();
+    this.visibilityUnlisten_();
+    this.visibilityUnlisten_ = null;
+    this.targetObserverMultimap_.clear();
+    this.targetDisplayMap_.clear();
+  }
+
+  /**
+   * @param {!Element} target
+   * @param {function(boolean)} callback
+   */
+  observe(target, callback) {
```

Comment:
not a particular actionable comment, but this file is remarkably similar to the new size-observer

Reply:
It's definitely turning into a pattern here. Not sure if we can canonize it in any way though.